File: python/heap-priority-queue/find-median-from-data-stream.py
import heapq


class MedianFinder:

    def __init__(self):
        self.above = []
        self.below = []
        self.count = 0

    def addNum(self, num: int) -> None:
        if self.count:
            below = -heapq.heappop(self.below)
            if self.count % 2 == 0:
                heapq.heappush(self.below, -below)
                above = heapq.heappop(self.above)
                if num > above:
                    heapq.heappush(self.below, -above)
                    heapq.heappush(self.above, num)
                else:
                    heapq.heappush(self.above, above)
                    heapq.heappush(self.below, -num)
            else:
                if num > below:
                    heapq.heappush(self.below, -below)
                    heapq.heappush(self.above, num)
                else:
                    heapq.heappush(self.above, below)
                    heapq.heappush(self.below, -num)
        else:
            heapq.heappush(self.below, -num)

        self.count += 1

    def findMedian(self) -> float:
        if self.count % 2 == 0:
            below = -heapq.heappop(self.below)
            above = heapq.heappop(self.above)
            heapq.heappush(self.above, above)
            median = (below + above) / 2
        else:
            below = -heapq.heappop(self.below)
            median = below

        heapq.heappush(self.below, -below)

        return median

# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: c++/heap-priority-queue/find-median-from-data-stream.cc
class MedianFinder {
public:
    MedianFinder() {
        count_ = 0;
    }
    
    void addNum(int num) {
        if (higher_.size() == lower_.size()) {
            higher_.push(-num);
            int low = -higher_.top();
            higher_.pop();
            lower_.push(low);
        } else {
            lower_.push(num);
            int high = lower_.top();
            lower_.pop();
            higher_.push(-high);
        }
        count_ += 1;
    }
    
    double findMedian() {
        if (count_ % 2 == 0) {
            return static_cast<double>(lower_.top() - higher_.top()) / 2.0;
        } else {
            return static_cast<double>(lower_.top());
        }
    }

private:
    std::priority_queue<int> higher_;
    std::priority_queue<int> lower_;

    int count_;
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */
 